### PR TITLE
Implement the new First Book authentication API

### PR DESCRIPTION
**Pull request by leonardr**
This branch implements the new First Book authentication API. It resolves https://jira.nypl.org/browse/SIMPLY-1719.

The new API is located at https://ebooksuat.firstbook.org/api/, but I assume "uat" stands for User Acceptance Testing and the final name will change. Instead of sending barcode and PIN directly in the URL to the API, we create a JWT which includes barcode and PIN in the payload, sign it with a shared secret, and send the JWT over in the URL. So the final URL we use looks like this:

https://ebooksuat.firstbook.org/api/eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJiYXJjb2RlIjoiQUJDREVGIiwicGluIjoiMzU3NCIsImlhdCI6IjE1NDk0OTE4MDkifQ.zJ1QXAJbQExrfbP227mNTm_qYri2PKNw_Fpo1qzBZbw

I made a copy of firstbook.py and test_firstbook.py, and modified it as necessary to accommodate the new API. I'll point out the major changes, but I think it's good to go over the whole thing since I don't think you're familiar with the  First Book API.

For now, the circulation manager will support both versions of the API. Once we get Open Ebooks using the new version, we can remove support for the old API.

### CODE REVIEW

File: api/firstbook2.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        return PatronData(
+            permanent_id=username,
+            authorization_identifier=username,
+        )
+
+    # End implementation of BasicAuthenticationProvider abstract methods.
+
+    def remote_pin_test(self, barcode, pin):
+        jwt = self.jwt(barcode, pin)
+        url = self.root + jwt
```
Comment (leonardr) [inline api/firstbook2.py]: This is new -- previously we stuck the barcode and PIN on to the end of `url`; now we call `jwt` and put _that_ on the end.

File: api/firstbook2.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        if response.status_code != 200:
+            msg = "Got unexpected response code %d. Content: %s" % (
+                response.status_code, response.content
+            )
+            raise RemoteInitiatedServerError(msg, self.NAME)
+        if self.SUCCESS_MESSAGE in response.content:
+            return True
+        return False
+
+    def jwt(self, barcode, pin):
```
Comment (leonardr) [inline api/firstbook2.py]: This whole method is new.

File: api/firstbook2.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+            # Simulate a bad connection.
+            raise requests.exceptions.ConnectionError("Could not connect!")
+        elif self.failure_status_code:
+            # Simulate a server returning an unexpected error code.
+            return MockFirstBookResponse(
+                self.failure_status_code, "Error %s" % self.failure_status_code
+            )
+        parsed = urlparse.urlparse(url)
+        token = parsed.path.split("/")[-1]
+        barcode, pin = self._decode(token)
```
Comment (leonardr) [inline api/firstbook2.py]: The `_decode` call is new, as is `_decode` itself. Previously there was nothing to decode, so we just looked for the barcode and pin in `self.valid`.

File: api/firstbook2.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+        # The barcode and pin must be present in self.valid.
+        if barcode in self.valid and self.valid[barcode] == pin:
+            return MockFirstBookResponse(200, self.SUCCESS)
+        else:
+            return MockFirstBookResponse(200, self.FAILURE)
+
+    def _decode(self, token):
```
Comment (leonardr) [inline api/firstbook2.py]: This whole thing is new.

File: tests/test_firstbook2.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        # Let's see what the mock API had to work with.
+        requested = self.api.request_urls.pop()
+        assert requested.startswith(self.api.root)
+        token = requested[len(self.api.root):]
+
+        # It's a JWT, with the provided barcode and PIN in the
+        # payload.
+        barcode, pin = self.api._decode(token)
+        eq_("ABCD", barcode)
+        eq_("1234", pin)
```
Comment (leonardr) [inline tests/test_firstbook2.py]: Previously we didn't bother testing which request was sent.

File: tests/test_firstbook2.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        os.environ['AUTOINITIALIZE'] = "False"
+        from api.app import app
+        self.app = app
+        del os.environ['AUTOINITIALIZE']
+        with self.app.test_request_context("/"):
+            doc = self.api.authentication_flow_document(self._db)
+            eq_(self.api.DISPLAY_NAME, doc['description'])
+            eq_(self.api.FLOW_TYPE, doc['type'])
+
+    def test_jwt(self):
```
Comment (leonardr) [inline tests/test_firstbook2.py]: This whole test is new.

File: api/firstbook2.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        # which is also the authorization identifier.
+        return PatronData(
+            permanent_id=username,
+            authorization_identifier=username,
+        )
+
+    # End implementation of BasicAuthenticationProvider abstract methods.
+
+    def remote_pin_test(self, barcode, pin):
+        jwt = self.jwt(barcode, pin)
```
Comment (EdwinGuzman) [inline api/firstbook2.py]: Where is this `self.jwt` function defined?

Comment (EdwinGuzman) [reply api/firstbook2.py]: Never mind, got slightly confused before.

**Pull request merged**